src/obfuscator.ts

```diff
@@ -0,0 +1,267 @@
+    { $project: projection}, // Limit to just obfuscateable fields
+    { $out: targetCollection.def.dbName } // Will overwrite what is currently in collection
+  ]
+  
+  const emptyCursor = await sourceCollection.db.aggregate(migrationSteps, options);
+
+  // $out does not take effect until the empty cursor is acted upon.  
+  // .toArray() should not return any data
+  // 'Works as designed' https://jira.mongodb.org/browse/NODE-1398
+  await emptyCursor.toArray();
+  Tyr.info(`Completed migration of obfuscateable fields from ${sourceCollection.name} to ${targetCollection.name}`);
+}
+
+// Probably not as efficient as mapreduce
+// TODO: recovery of mid encryption failure
+// TODO: trace logging
+Tyr.encryptCollection = async (collection: Tyr.CollectionInstance, masterKey: string) => {
+  await encryptDecryptData(collection, masterKey, {}, false);
+}
+
+const encryptDecryptData = async (collection: Tyr.CollectionInstance, masterKey: string, query: Tyr.MongoQuery, decrypt?: boolean) => {
+
+  const aesUtil = new AesUtil(masterKey);
+  const cursor = await collection.db.find(query);
+  let bulkOp = collection.db.initializeUnorderedBulkOp();
+
+  const count = await cursor.count();
+  let doc;
+  for (let i = 1; i <= count; i++) {
+    doc = await cursor.next();
+
+    bulkOp.find({ _id: doc._id }).update({ $set: encryptDocument(doc, aesUtil, decrypt) });
+
+    if ((i % defaultBatchSize === 0) || i === count) {
+      Tyr.info('Flushing encrypted batch');
+      const result: BulkWriteResult = await bulkOp.execute();
+
+      if (!result.ok) {
+        throw new Error(`Encryption of collection failed`)
+      }
+      bulkOp = collection.db.initializeUnorderedBulkOp();
+    }
+  }
+}
+// TODO: Need to think of wording
+const encryptDocument = (doc: any, aesUtil: AesUtil, decrypt?: boolean): object => {
+  //TODO: figure out better approach to typing obj
+  let encObj: any = {};
+
+  Object.getOwnPropertyNames(doc).forEach(key => {
+    if (key !== "_id") {
+      //Encryption util is simplistic right now
+      //should replace with established library
+      const val = doc[key];
+      //simple attempt at supporting datatypes other than string
+      encObj[key] = decrypt ? aesUtil.decrypt(doc[key]) : aesUtil.encryptString((typeof val === 'string') ? val : JSON.stringify(val));
+    }
+  });
+
+  return encObj;
+}
+
+Tyr.restoreObfuscatedData = async (targetCollection: Tyr.CollectionInstance, sourceCollection: Tyr.CollectionInstance, query ?: Tyr.MongoQuery, decryptionKey ?: string) => {
+  //decrypt collection
+  if (decryptionKey) {
+    // need better method naming
+    await encryptDecryptData(sourceCollection, decryptionKey, query, true);
+  }
+  //migrate data back to targetCollection
+  await migrateData(targetCollection, sourceCollection, query);
+
+}
+
+const migrateData = async (targetCollection: Tyr.CollectionInstance, sourceCollection: Tyr.CollectionInstance, query?: Tyr.MongoQuery) => {
+  const q = query ? query : {};
```

Comment:
Non-blocking: You can do [default paramaters](https://www.typescriptlang.org/docs/handbook/functions.html#optional-and-default-parameters) if you like using this syntax:
```
async (..., query: Tyr.MongoQuery = {})
```

Reply:
interesting, I just assumed the function signature had to match the interface